### DISCUSSION TITLE
[Chapter 05] @InitBinder

### DISCUSSION THREAD

**Discussion by BvrPark** [cowaine/spring-boot-for-msa#33] [Chapter 05. REST APIs with Spring MVC] [answered]
Q1-1
p.261에서 @InitBinder에 대한 예시가 나왔는데요. 
그럼 저 Controller에서는 다른 메서드보다 우선 initBinder메서드가 먼저 실행되서 유효성 검사를 하는건가요...?

Q1-2
1-1이 맞으면 @InitBinder가 Controller에 있는 공통적인 유효성 검증 작업(?)을 한다고 생각이 드는데
한 Controller 클래스 내에서 메서드 마다 다른 유효성 검증이 필요할 때에는 @InitBinder보다 @Valid를 사용하는 것이 맞는거겠죠?

**Comment by Irisation23** [comment, marked as answer]
1-1 에 대한 질문 의도가 `initBinder()` 먼저 실행되어서 유효성 검증을 이어 나가는 것이 아니라 유효성 검증을 위한 또 다른 방법이라 생각 됩니다.

먼저 책에서 예제로 든 상황을 가정해 봅시다.

1. 체크 아웃 가능 시간이 체크 인 가능 시간보다 앞설수 없다. e.g) 체크 아웃 = 2023-08-24 , 체크 인 = 2023-08-25

해당 검증을 `JSR-303`에 알맞게 할 수 없기 때문에, 위의 검증 로직을 처리할 수 있는 `HotelRoomReserveValidator` 를 바인더에 초기화 하는 과정이 필요합니다.

---

1-2는 올바른 가정이 아니라 생각되고, 위에서 답변 한 대로 필요한 상황에 알맞은 예외 처리가 필요하다고 생각됩니다.

명확하게 `@InitBinder` 로 바인딩 처리 클래스를 지정했다 해도 `@Valid`는 반드시 작성되어야 올바른 예외처리를 할 수 있습니다.

**Comment by Youngju-Jang** [comment]
근데, 저는 지금 262페이지 상단예시대로 요청보냈을때 파라미터단에서 바로 예외처리 걸리지않고,
아래 서비스단의 예외가 나오는 상황입니다 

서비스로 넘어가기전에 잡힐거라 생각했는데, 넘어간다는건 순서가 InitData가 먼저가 아닌건지도..?, 아니면 코드에 뭔가 빠졌을까요

![image](https://github.com/cowaine/spring-boot-for-msa/assets/77484084/9f063dee-8f65-43ff-aae5-98ea9992e0bc)

`validator` 코드는 [여기](https://github.com/cowaine/spring-boot-for-msa/blob/feature%2Fyoungju-jang/youngju-jang/src/main/java/com/cowaine/youngjujang/ch5/domain/utils/validator/HotelRoomReserveValidator.java#L1) 있어요

**Reply by Youngju-Jang** [reply]
죄송해요, 오류나던 서비스단 주석처리하고 실행해보니 제코드에선 `@InitBinder`관련부분이 아예 걸리지가 않네요;;

**Comment by JoisFe** [comment]
cc. @Youngju-Jang 

- 정상 요청을 해본 경우
<img width="461" alt="image" src="https://github.com/cowaine/spring-boot-for-msa/assets/90208100/dfe5e37a-71f8-45cf-8dbd-7d73d94f4e43">

- checkInDate 에 값을 넣지 않은 경우
<img width="468" alt="image" src="https://github.com/cowaine/spring-boot-for-msa/assets/90208100/ca08e32b-fa26-44a2-be7e-51ad58dc1287">

<img width="1257" alt="image" src="https://github.com/cowaine/spring-boot-for-msa/assets/90208100/6f26b714-afae-4a19-836e-a7ec40242fda">

이렇게 잡히는데 이 부분이 안잡힌다는 건가요?

디버깅 모드로
<img width="855" alt="image" src="https://github.com/cowaine/spring-boot-for-msa/assets/90208100/6c3520eb-0f3d-4525-92d9-800c5892fc9f">

컨트롤러에서 해당 부분 보다
<img width="983" alt="image" src="https://github.com/cowaine/spring-boot-for-msa/assets/90208100/f0bc1f90-4938-400b-a862-81423605c35c">

HotelRoomReserveValidator 의 validate 메서드에 먼저 걸리는 것을 확인할 수 있습니다.

**Reply by Youngju-Jang** [reply]
오맞아요 디버깅했을때 가는것은확인했는데, 저는 저기서 걸린다음에 에러에 rejectValue에 뭐가 들어갔으니까 알아서 에러가 발생할거라고 생각했었는데,
따로 에러가발생을 안하더라고, 위에 올려주신 두번째처럼 그냥 결과값이 나오더라구요..??
